store.go

```diff
@@ -2953,6 +2977,32 @@ func (s *store) ApplyDiffFromStagingDirectory(to, stagingDirectory string, diffO
 	return err
 }
 
+func (s *store) PutLayerFromStagingDirectory(args PutLayerFromStagingDirectoryOptions) (*Layer, error) {
+	rlstore, rlstores, err := s.bothLayerStoreKinds()
+	if err != nil {
+		return nil, err
+	}
+	if err := rlstore.startWriting(); err != nil {
+		return nil, err
+	}
+	defer rlstore.stopWriting()
+	if err := s.containerStore.startWriting(); err != nil {
+		return nil, err
+	}
+	defer s.containerStore.stopWriting()
+
+	layer, _, err := s.putLayerLocked(rlstore, rlstores, args.ID, args.ParentLayer, args.Names, args.MountLabel, args.Writeable, args.LayerOptions, nil)
+	if err != nil {
+		return layer, err
+	}
+
+	if err := rlstore.ApplyDiffFromStagingDirectory(layer.ID, args.StagingDirectory, args.DiffOutput, args.DiffOptions); err != nil {
+		rlstore.Delete(layer.ID)
```

Comment:
Please either return both errors or add a warning-level log.

Otherwise, logs from production where some layers are missing for an unknown reason are very difficult to debug. (Sure, ideally the caller should provide a unique ID that is not reused in the meantime… but if we are looking at logs, there is a bug involved. And, yes, c/image didn’t have that.)